exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ScanBatch.java

```diff
@@ -152,97 +157,75 @@ public void kill(boolean sendUpstream) {
     }
   }
 
-  private void releaseAssets() {
-    container.zeroVectors();
-  }
-
-  private void clearFieldVectorMap() {
-    for (final ValueVector v : mutator.fieldVectorMap().values()) {
-      v.clear();
-    }
-  }
-
   @Override
   public IterOutcome next() {
     if (done) {
       return IterOutcome.NONE;
     }
     oContext.getStats().startProcessing();
     try {
-      try {
-        injector.injectChecked(context.getExecutionControls(), "next-allocate", OutOfMemoryException.class);
-
-        currentReader.allocate(mutator.fieldVectorMap());
-      } catch (OutOfMemoryException e) {
-        clearFieldVectorMap();
-        throw UserException.memoryError(e).build(logger);
-      }
-      while ((recordCount = currentReader.next()) == 0) {
+      while (true) {
         try {
-          if (!readers.hasNext()) {
-            // We're on the last reader, and it has no (more) rows.
-            currentReader.close();
-            releaseAssets();
-            done = true;  // have any future call to next() return NONE
-
-            if (mutator.isNewSchema()) {
-              // This last reader has a new schema (e.g., we have a zero-row
-              // file or other source).  (Note that some sources have a non-
-              // null/non-trivial schema even when there are no rows.)
+          injector.injectChecked(context.getExecutionControls(), "next-allocate", OutOfMemoryException.class);
```

Comment:
From the PR description:

> 1. Skip RecordReader if it returns 0 row && present same schema. A new schema (by calling Mutator.isNewSchema() ) means either a new top level field is added, or a field in a nested field is added, or an existing field type is changed.

The code, however, adds an additional condition: if implicit fields change. (But, as noted below, that should never occur in practice.)

What happens on the first reader? There is no schema, so *any* schema is a new schema. Suppose the file is JSON and the schema is built on the fly. Does the code handle the case that we have no schema (first reader), and that reader adds no columns?

Or, according to the logic that the downstream wants to know the schema, even if there are no records, do we send an empty schema (schema with no columns) downstream, because that is an accurate representation of an empty JSON file?

What happens in the case of an empty JSON file following a non-empty file? In this case, do we consider the empty schema as a schema change relative to a non-empty change?

In short, can we generalize this first rule a bit?

> 2. Implicit columns are added and populated only when the input is not empty, i.e. the batch contains > 0 row or rowCount == 0 && new schema.

How does this interact with a scan batch that has only one file, and that file is empty? Would we return the empty schema downstream? With the implicit columns?

> 3. ScanBatch will return NONE directly (called as "fast NONE"), if all its RecordReaders haver empty input and thus are skipped, in stead of returing OK_NEW_SCHEMA first.


This is just a bit ambiguous. If the reader is JSON, then an empty file has an empty schema (for reasons cited above.)

But, if the input is CSV, then we *always* have a schema. If the file has column headers, then we know that the schema is, say, (a, b, c) because those are the headers. Or, if the file has no headers, the schema is always the `columns` array. So, should we send that schema downstream? If so, should it include the implicit columns?

This, in fact, raises another issue (out of scope for this PR): if we return an empty batch with non-empty schema, we have no place to attach the implicit columns that will allow the user to figure out that, say, "foo.csv" is empty.

On the other hand, if we say that an empty CSV file has no schema, then we can skip that file. The same might be true of JSON. What about Parquet? We'd have a schema even if there are no rows. Same with JDBC. Should we return this schema, even if the data is empty?

Finally, do we need special handling for "null" files: a file that no longer exists on disk and so has a completely undefined schema? An undefined schema is different than an empty schema. Undefined = "we just don't know what the schema might be." Empty = "we know there is a schema and that schema is empty." A missing CSV file has an undefined schema. We could argue that a 0-length (or, a one-byte file with the only content being a newline) is a valid file, but it is a file with no columns, hence an empty schema.

Or, should we have a rule that says something like:

1. If a file is missing, or has an undefined schema, skip it.
2. If a file has a schema (empty or not) and no rows, then:
  * If this is the first file, remember the schema and skip to the next.
  * If this file follows a non-empty file, trigger a schema change.
4. If a file has a rows, then:
  * If the previous schema was empty, ignore that empty schema.
  * If the previous schema was non-empty, and different, return OK_NEW_SCHEMA
  * If the previous schema was the same, return OK.
5. At EOF:
  * If the previous schema was empty, and the previous file had no rows, return that schema.
  * If the previous schema was empty, or no non-empty files, return NONE.

We can reduce these to a simpler set of rules:

* We ignore all empty or missing schemas.
* We ignore empty files unless they are the only file(s) or they have a schema different than the next non-empty file.

Other rules are possible. One is simply to ignore all 0-length files regardless of whether the schema is the same or different than surrounding files. If all files return 0 records, just return NONE. (These are the rules adopted, as of now, by the new scan operator; though we will change them to match the rules here one we pin them down.)

Reply:
This patch tries to decouple the logic of record reader and scanbatch:
 - Record reader is responsible to add vectors to batch (via Mutator), and populate data
 - ScanBatch is responsible to interpret the output of record reader, by checking rowCount && Mutator.isNewSchema() to decide whether return OK_NEW_SCHEMA, OK, or NONE. 

> What happens on the first reader? There is no schema, so any schema is a new schema. Suppose the file is JSON and the schema is built on the fly. Does the code handle the case that we have no schema (first reader), and that reader adds no columns?

It's not true "any schema is a new schema". If the first reader has no schema and adds no columns, then Mutator.isNewSchema() should return false. Mutator.isNewSchema() returns true only after the last call, one or more happens

- a new top level field is added, 
- a field in a nested field is added, 
- an existing field type is changed

You may argue a more appropriate way to represent an empty JSON file is an empty schema. However, such idea would lead to various schema conflicts in down-stream operator, if other scan thread has non-empty JSON files. This is exactly what happened before this patch. 

The proposal in this patch is to **ignore** empty JSON, since 1)rowCount=0, 2)no new column were added to batch. 
 - If all the record readers for a scan thread return with rowCount = 0, and produce no new schema, then this Scan thread should return 'NONE' directly, without returning OK_NEW_SCHEMA. 
 - If at least one of reader return either with >0 row, or new schema, then Scan thread will return batch with new schema.
- If all scan threads returns 'NONE' directly, implying the entire table does not have data/schema, this is what Project.handleFastNone() will deal with.

>But, if the input is CSV, then we always have a schema. If the file has column headers, then we know that the schema is, say, (a, b, c) because those are the headers. Or, if the file has no headers, the schema is always the columns array. So, should we send that schema downstream? If so, should it include the implicit columns?

If CSV always adds columns (either _a,b,c, or columns_), then ScanBatch will produce a batch with (a, b, c), or columns. It does not make sense to ignore those schema.  
  - In the case of file with header,  file with _a,b,c_ will lead to a batch with (a,b,c) while a file with _a,b,c,d_ will lead to a batch with (a,b,c,d). Those two files will cause a schema change, which is expected behavior.
  - In the case of file without header, all files will produce a batch with columns, which means there would be no schema change across different files, regardless whether they have row=0, or row > 0.